packages/perseus/src/util/parse-perseus-json/general-purpose-parsers/array.test.ts

```diff
@@ -0,0 +1,82 @@
+import {failure, success} from "../result";
+
+import {array} from "./array";
+import {string} from "./string";
+import {anyFailure, ctx} from "./test-helpers";
+
+describe("array", () => {
+    const arrayOfStrings = array(string);
+
+    it("accepts an empty array", () => {
+        expect(arrayOfStrings([], ctx())).toEqual(success([]));
+    });
+
+    it("accepts a array of one element", () => {
+        expect(arrayOfStrings(["hi"], ctx())).toEqual(success(["hi"]));
+    });
+
+    it("accepts a array of multiple elements", () => {
+        expect(arrayOfStrings(["a", "b", "c"], ctx())).toEqual(
+            success(["a", "b", "c"]),
+        );
+    });
+
+    it("rejects a string", () => {
+        expect(arrayOfStrings("blah", ctx())).toEqual(anyFailure);
+    });
+
+    it("rejects a array with one element of the wrong type", () => {
+        expect(arrayOfStrings([99], ctx())).toEqual(anyFailure);
+    });
+
+    it("rejects a array with multiple elements if any has the wrong type", () => {
+        expect(arrayOfStrings(["ok", 99, "ok"], ctx())).toEqual(anyFailure);
+    });
+
+    it("indicates the bad element in the failure message", () => {
+        const theArray = [
+            "ok",
+            "ok",
+            99, // index 2
+        ];
+
+        const result = arrayOfStrings(theArray, ctx());
+
+        expect(result).toEqual(
+            failure(
+                expect.objectContaining({
+                    message: "expected string, but got 99",
+                    path: [2],
+                }),
+            ),
+        );
```

Comment:
This expected value can be abbreviated `parseFailureWith({message: ..., path: ...})`